### ISSUE TITLE
threshold_adaptive all black

### ISSUE THREAD

**Issue by opyate** [scikit-image/scikit-image#1403]
I saw this with the [previous version as commented on here](https://github.com/scikit-image/scikit-image/commit/e84abb1b67586d3a3620ce6a21e3314527083218#commitcomment-9990783), but now also with version ```0.11.0```.

I tried various combinations of ```blocksize``` and ```offset```:

    from skimage import io
    from skimage.filters import threshold_adaptive
    import numpy as np

    image = io.imread('input.png', as_grey=True) # e.g. http://i.imgur.com/PqykmyV.png

    for blocksize in np.arange(1, 41, 2):
        for offset in np.arange(0., 10., 0.1):
            binary = threshold_adaptive(image, blocksize, offset=offset)
            io.imsave('bin/binarised.%i.%f.png' % (blocksize,offset), binary)

...and they all come out black, albeit they have sizes ranging from 328 bytes to 3659 bytes, with the input 2864 bytes.

Am I doing something obviously wrong?

I'm running all this in an ```pyenv``` virtual environment.

    $ uname -a

    Linux redacted 3.13.0-46-generic #77~precise1-Ubuntu SMP \
        Mon Mar 2 23:24:50 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

    $ python --version
    Python 3.4.2

    $ pip freeze

    Cython==0.22
    decorator==3.4.0
    matplotlib==1.4.3
    networkx==1.9.1
    nose==1.3.4
    numpy==1.9.2
    Pillow==2.7.0
    pyparsing==2.0.3
    python-dateutil==2.4.0
    pytz==2014.10
    scikit-image==0.11.0
    scikit-learn==0.15.2
    scipy==0.15.1
    six==1.9.0

**Comment by opyate** [comment]
Hmmm, looks fine in ```Jupyter``` (installed in same virtualenv), so no idea what's wrong, really.

![iPython](http://i.imgur.com/GqfSfHN.png)

**Comment by opyate** [comment]
Never mind - I'll just do all my experiments in a notebook, since it seems to be rendering fine there.

**Comment by jni** [comment]
@opyate my guess is that they are being saved as uint8 images with values 0 and 1. Both of these will appear black when opened in standard image viewers. However, matplotlib will automatically stretch the values to the full luminance range, which is why it is visible.

One option is to save `binary.astype(np.uint8) * 255` instead of just `binary`.

**Comment by blink1073** [comment]
I'd like to take a look at this and see if we need to change the imsace behaviour.

**Comment by jni** [comment]
@blink1073 I think imsave is doing exactly what it's supposed to. uint8(True) -> 1. I despise putting in a binary image and getting back a {0, 255} image on load.

**Comment by opyate** [comment]
> my guess is that they are being saved as uint8 images with values 0 and 1. 

@jni thanks for the explanation.

**Comment by blink1073** [comment]
@jni, perhaps we should give a warning when imsave is about to save an image with low contrast?  That way the user would at least know what happened.

**Comment by jni** [comment]
@blink1073 sure. This goes to the same discussion as the `io.imshow`: what is "low contrast"? This is definitely something I want to do, but we need to answer that question well. (Remember that contrast is lower for a given step size at the dark end of the scale, ie 0-1 is less distinguishable than 254-255.)

**Comment by stefanv** [comment]
Catch the most common case and proceed from there!

**Comment by blink1073** [comment]
I would simply define low contrast as the ration of min/max of the image vs. the min/max of the dtype range being less than some threshold, say 5%.

**Comment by blink1073** [comment]
Or you could use the 1st and 99th percentiles instead of min/max

**Comment by JDWarner** [comment]
+1 for extreme percentile checks.

**Comment by blink1073** [comment]
How about a utility function:

```python

from skimage import dtype_limits
import numpy as np
import warnings

def check_contrast(image, threshold_percent=5):
      dlimits = dtype_limits(image)
      limits = np.percentile(image, [1, 99])
      ratio = float(limits[1] - limits[0]) / float(dlimits[1] - dlimits[0])

      if ratio < threshold_percent:
          warnings.warn('Image contrast is very low (%0.1f%% of usable range)' % ratio))
```

**Comment by stefanv** [comment]
It is so annoying that warnings only get issued once by default.  I
guess it's the best we have though.  I like your implementation.

**Comment by jni** [comment]
@blink1073 I'd make the following changes:

- change the default `threshold_percent` to 0.05, or you'll always be warning. ;)
- call it something more descriptive, such as `min_contrast` or `mean_data_range_ratio`.
- add an option for the percentile stretching, and, I would say, make it 0 by default. (I don't feel strongly about the default, but I do about the user's ability to change this.)
- return a bool, rather than warn. Functions that use it can decide whether to warn (e.g. `io.imsave` would do this) or alter the image display (e.g. `io.imshow` would do this).

Other than that, I like having this function because if later one of us decides to inquire more deeply into that perceptual nonlinearity, we can just update the function.

**Comment by blink1073** [comment]
Fixed by #1412.
